### PR TITLE
NN-2251 return new dto containing absent and offender details to drive stats screen

### PR DESCRIPTION
**Pull request by JamieDraperUK**

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/Elite2ApiService.java [hunk 1]
```diff
@@ -68,7 +68,8 @@ public String getOffenderNoFromBookingId(final Long bookingId) {
     public List<Long> getBookingIdsForScheduleActivitiesByDateRange(final String prisonId, final TimePeriod period, final LocalDate fromDate, final LocalDate toDate) {
         final var url = "/schedules/{prisonId}/activities-by-date-range?fromDate={fromDate}&toDate={toDate}&timeSlot={period}";
 
-        final var responseType = new ParameterizedTypeReference<List<Map>>() {};
+        final var responseType = new ParameterizedTypeReference<List<Map>>() {
+        };
```
Comment (Nimphal) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/Elite2ApiService.java]: Definitely not a blocker but looks like your IDE has been a bit overzealous again.

Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/Elite2ApiService.java]: I need to look into it but I suspect that's the standard Java way to format code with curly braces.

Comment (brightonsbox) [reply src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/Elite2ApiService.java]: Concise braces in Java are usually considered acceptable: https://google.github.io/styleguide/javaguide.html#s4.1.3-braces-empty-blocks

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/dto/AttendancesResponse.kt [hunk 2]
```diff
@@ -13,6 +13,7 @@ import lombok.NoArgsConstructor
 @Builder(toBuilder = true)
 @ApiModel(description = "Attendances response")
 data class AttendancesResponse(@ApiModelProperty(required = true, value = "Set of attendances")
-                               var attendances: Set<AttendanceDto>? = null
+                               var attendances: Set<AttendanceDto>? = null,
+                               var absences: Set<AbsenceDto>? = null
```
Comment (brightonsbox) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/dto/AttendancesResponse.kt]: Add another @ApiModelProperty for the absences variable?

File: src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/AttendanceService.java [hunk 3]
```diff
@@ -276,32 +276,69 @@ private Boolean isAttendanceLocked(final Attendance attendance) {
...
+                details.getComment(),
+                attendance.getComments(),
+                details.getCellLocation(),
+                details.getFirstName(),
+                details.getLastName()
+        );
+    }
+
+    private Predicate<? super OffenderDetails> findAttendance(final Long bookingId, final Long eventId, final TimePeriod period) {
+        return offender -> offender.getBookingId().equals(bookingId) &&
```
Comment (brightonsbox) [inline src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/AttendanceService.java]: Probably minor, but it looks like OffenderDetails has a nullable bookingId? Could this raise a NullPointerException if it's not there? Same with the eventId below.

Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/AttendanceService.java]: Good spot, bookingId is fine, however eventId, at least in dev is sometimes null due to crappy data. I'll fix it anyways just to be safe.

**Pull request merged**